### PR TITLE
Fix/1641 stt code in user

### LINKED ISSUE #1641: As a sys admin, I need to see a meaningful STT name and FIPS/TRIBE code in a user's profile
**Description:**
After #1387 is merged, the STT field is no longer available in the user profile (See [below](#Supporting-Documentation)). Instead we have a `location_id` (which maps to the id in the STT model) which is not a meaningful value for staff responsible for ensuring that STT users (i.e. `data analysts`) can only submit data on behalf of their own STT. STT location and STT code are more meaningful for user management purposes.

These fields can be **read_only** at least until such a time as we iron out the **change** request flow. 

**Acceptance Criteria:** 

- [x] Allow saving of user STT with or without Role
- [x] STT_CODE code and needs to accept 3 digits rather than 2 (current). Leading zeros should be maintained (e.g. 078) (req removed from #217)
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

AC to be handled in separate PR
- [x] update the django admin user details page to make the relationship between a user and the STT they are associated with more obvious
   - dropdown select list for stt (if stt user, deactivated if regional user)
   - [STT Name](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/stts/management/commands/data/states.csv) field (e.g. Virginia) appears in user profile 
   - [STT_code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/stts/management/commands/data/states.csv) field (e.g. 11) appears as location ID in user profile
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [ ] _design an updated mockup for the user profile_
- [ ] _add fields changes to user profile_ 
- [ ] change filtering of STT objects to be based on STT code

#### Supporting Documentation:

- _Current user profile fields as of v2.2.2_
1641p1

**Notes:**
- #217 blocks this because the `STT_CODE` field is not in the object
- Make new form validation permissive and easy to use

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #1641
* update the django admin user details page to make the relationship between a user and the STT they are associated with more obvious 
   * dropdown select list for stt (if stt user, deactivated if regional user)
   * [STT Name](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/stts/management/commands/data/states.csv) field (e.g. Virginia) appears in user profile
   * [STT_code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/stts/management/commands/data/states.csv) field (e.g. 11) appears as location ID in user profile

This change removes the `location_type` and `location_id` fields managed by the `location` `GenericForeignKey` in favor of separate `region` and `stt` fields on the `user` model. This reduces complexity some - the major trickiness is around validation - and allows us to simplify the admin experience of selecting an stt location. The django admin User page was updated to allow an `stt` to both display and be searched by `name` or `stt_code`.

## How to Test
```
cd tdrs-frontend && npm run start
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Request access via the form, select a location.
3. In the Django Admin, edit the user and verify that the selected location matches what was entered in the Request Access form.
4. Select the Data Analyst role, select the "Pending" approval status, and save the user. Verify on the frontend that the "Profile" tab becomes visible and the "Data Files" tab is still not accessible. The "Profile" tab should show the name of the user's assigned location.
5. In the Django Admin, select the "Approved" approval status for the user, verify the "Data Files" tab becomes visible on the frontend.
6. Navigate to the Data Files tab, submit a data file and verify that the STT associated with the data file matches the user's STT.
7. In the Django Admin, select a new location for the user.
8. In the frontend, navigate to the Data Files tab and submit a new data file. Verify the associated STT matches the user's new location.
9. In the Django Admin, select the "OFA Regional Staff" group and remove the user's STT assignment. Save the user (do not add a Region yet\*).
10. Now add a region, and save the user. Verify that the user's new location appears in the "Profile" tab on the frontend.
11. Navigate to the "Data Files" tab. Verify a message is shown stating that no STT is assigned to the user.

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] update the django admin user details page to make the relationship between a user and the STT they are associated with more obvious 
   * dropdown select list for stt (if stt user, deactivated if regional user)
   * [STT Name](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/stts/management/commands/data/states.csv) field (e.g. Virginia) appears in user profile
   * [STT_code](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/tdpservice/stts/management/commands/data/states.csv) field (e.g. 11) appears as location ID in user profile
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/users/test/test_models.py
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/test/test_models.py:19]: this is one of my unhappy moments - i wasn't able to perfectly preserve the `save` behavior for a weird mess of reasons.

django `ForeignKey` fields can't have `setter` overrides, which is what previously provided this validation. so, I moved the validation to the `User.validate_location` method, which is called by `User.clean`. This covers the admin page throwing validation errors on save. Since DRF's `ModelSerializer` _does not_ call the model's `clean` method when validating or saving, I had to override the `ModelSerializer.update` method to call it.

I tried adding the `User.validate_location` call to the `User.save` override, which would have made a change to the serializer unnecessary, but the django admin will show the ugly django 500 error page instead of the pretty form-level error message. If that's acceptable, I'm happy to change it, otherwise we have to manually validate users before we save them. 

This is ultimately a patterning issue which can be addressed by the use of `ModelManagers` to handle all state updates and interactions with the model class
* https://www.dabapps.com/insights/django-models-and-encapsulation/

but I thought it might be overkill.

File: tdrs-backend/tdpservice/users/admin.py
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/admin.py:46]: this logic has all been moved to the `User.validate_location` method. Actually, it was already there.

File: tdrs-backend/tdpservice/users/models.py
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/models.py:179]: again, this logic was effectively duplicated for the individual actions of `ModelForm` saves, `User.clean` validation, and (here) setting fields directly. Getting rid of the `GenericForeignKey` in favor of individual fields lets us consolidate all the logic with no change to how the model is used.

File: tdrs-backend/tdpservice/users/admin.py
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/admin.py:39]: I thought this would be a nice add - it makes the `stt` dropdown select in the django admin searchable by the `STTAdmin.search_fields` values (`name` and `stt_code`).

File: tdrs-backend/tdpservice/users/migrations/0037_auto_20230306_1434.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/migrations/0037_auto_20230306_1434.py]: Shouldn't we log instead of print?

File: tdrs-backend/tdpservice/users/models.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/models.py:47]: If we are not using limit, we should delete it

File: tdrs-backend/tdpservice/users/models.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/models.py:47]: What's the reason for setting it to null instead of cascading deletion?

File: tdrs-backend/tdpservice/users/models.py
Comment (jtimpe) [reply tdrs-backend/tdpservice/users/models.py:47]: this is deleted code

File: tdrs-backend/tdpservice/users/models.py
Comment (jtimpe) [reply tdrs-backend/tdpservice/users/models.py:47]: When deleting a "state" STT, a cascading delete would delete any "tribe" stts that were associated. I made the decision to set to `NULL` for the `stt` field thinking since it was a relation to the same model, it might save a headache. For Region, I just copied the same pattern. Open to discussing this further

File: tdrs-backend/tdpservice/users/migrations/0037_auto_20230306_1434.py
Comment (jtimpe) [reply tdrs-backend/tdpservice/users/migrations/0037_auto_20230306_1434.py]: ....maybe. good catch, thank you.

File: tdrs-backend/tdpservice/users/models.py
Comment (raftmsohani) [reply tdrs-backend/tdpservice/users/models.py:47]: I meant to delete line 37

Comment (jtimpe) [reply tdrs-backend/tdpservice/users/models.py:47]: ah, i see. good catch, i'll get it removed. thanks!

**Review by raftmsohani — COMMENTED** [review]
Besides the two comments, it looks good to me!

**Review by raftmsohani — APPROVED** [review]
LGTM! Reviewed and comments are resolved

File: tdrs-backend/tdpservice/users/test/test_models.py
Comment (andrew-jameson) [reply tdrs-backend/tdpservice/users/test/test_models.py:19]: Hmm, noted, need to finish reading/comprehending this link sent but yes I recall in the past we've had this save/clean duality b/c of the implicit signals Django is sending around with models.

**Review by andrew-jameson — APPROVED** [review]

**Review by ADPennington — APPROVED** [review]
thanks @jtimpe 😄 this is a significant improvement for the sys admin experience. at some point in the future, we should figure out how to save a few more steps here (e.g. not having to use the pending status), but this is good enough to get us to the parity stage. 

@ttran-hub steps to assign new users a role to ensure they get properly configured approval email are below. I will also update the documentation in the DIGIT repo, for future reference. cc: @lfrohlich 

_**before this PR**_
From the user's profile:
1. select group: Data Analyst
2. change location type to: --------
3. change account approval status to: Pending
4. click: save and continue editing
5. change location type to: stt|stt
6. change account approval status to: Approved
7. click: Save

_**after this PR is merged**_
From the user's profile:
0. verify that the STT requested in the user profile is set and matches what is on the tracker)
1. select group: Data Analyst
2. change account approval status to: Pending
3. click: save and continue editing
4. change account approval status to: Approved
5. click: Save
**Note: do not set the `region` (this is needed for regional staff permissions only)**

**Pull request merged**
